code_annotations/extensions/base.py

```diff
@@ -19,14 +26,219 @@ def __init__(self, config, echo):
+        self.query = self.annotation_regex.format('|'.join(self.annotation_regexes))
+
+        self.ECHO.echo_v("{} extension regex query: {}".format(self.extension_name, self.query))
+
+    def _add_annotation_token(self, annotation):
+        """
+        Add annotations to our local lists during configuration.
+
+        Args:
+            annotation: An annotation token (ex. ".. pii::")
+
+        Raises:
+            TypeError
+        """
+        self.annotation_tokens.append(annotation)
+        self.annotation_regexes.append(re.escape(annotation))
+
+    def _add_annotation_group(self, annotation_group):
+        """
+        Add an annotation group to our local lists during configuration.
+
+        Args:
+            annotation_group: An annotation group from the configuration file.
+
+        Raises:
+            TypeError
+        """
+        for annotation in annotation_group:
+            if isinstance(annotation, six.string_types):
+                self._add_annotation_token(annotation)
+            elif isinstance(annotation, dict):
+                annotation_name = next(iter(annotation))
+                self._add_annotation_token(annotation_name)
+            else:
+                raise TypeError(
+                    '{} is an unknown type. Annotations must be strings or list/tuples.'.format(annotation_group)
+                )
+
+    def validate(self, file_handle):
+        """
+        Validate that any annotations in the given file are properly formatted.
+
+        Args:
+            file_handle: Open file handle for the file to validate, set to beginning of the file
+
+        Returns:
+            Tuple of (success, list strings describing reasons for failure)
+        """
+        return True
+
+    def search(self, file_handle):
+        """
+        Search for annotations in the given file.
+
+        Args:
+            file_handle: Handle for the file to validate
+
+        Returns:
+            List of dicts describing found annotations, or an empty list
+        """
+        txt = file_handle.read()
+
+        found_annotations = []
+
+        # Fast out if no annotations exist in the file
+        if any(anno in txt for anno in self.annotation_tokens):
+            for match in re.finditer(self.comment_regex, txt):
+                # Should only be one match
+                comment_content = [item for item in match.groups() if item is not None][0]
+                for inner_match in re.finditer(self.query, comment_content):
+                    # No matter how long the regex is, there should only be 2 non-None items,
+                    # with the first being the annotation token and the 2nd being the comment.
+                    cleaned_groups = [item for item in inner_match.groups() if item is not None]
+
+                    if len(cleaned_groups) != 2:  # pragma: no cover
+                        raise Exception('Number of found items in the list is not 2, something is very wrong.')
```

Comment:
Nit: On the scale of wrong things, is this `very`? 😺 Maybe adding a print of the cleaned_groups *or* the number of found items would be better?

Reply:
I'd classify it as very bad, it means the whole regex is broken and nothing being reported can be trusted, but I also agree that better output is required!